lib/demo/accounts/user.ex

```diff
@@ -0,0 +1,116 @@
+defmodule Demo.Accounts.User do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  schema "users" do
+    field :email, :string
+    field :password, :string, virtual: true
+    field :encrypted_password, :string
+    field :confirmed_at, :naive_datetime
+
+    timestamps()
+  end
+
+  @doc """
+  A user changeset for registration.
+
+  It is important to validate the length of both e-mail and password.
+  Otherwise databases may truncate them without warnings, which could
+  lead to unpredictable or insecure behaviour. Long passwords may also
+  be very expensive to encrypt.
+  """
+  def registration_changeset(user, attrs) do
+    user
+    |> cast(attrs, [:email, :password])
+    |> validate_email()
+    |> validate_password()
+  end
+
+  defp validate_email(changeset) do
+    changeset
+    |> validate_required([:email])
+    |> validate_format(:email, ~r/^[^\s]+@[^\s]+$/, message: "must have the @ sign and no spaces")
+    |> validate_length(:email, max: 160)
+    |> unsafe_validate_unique(:email, Demo.Repo)
+    |> unique_constraint(:email)
+  end
+
+  defp validate_password(changeset) do
+    changeset
+    |> validate_required([:password])
+    |> validate_length(:password, min: 12, max: 80)
```

Comment:
What about something like...

```
validate_format(:password, ~r/(?=.*[A-Z])(?=.*[a-z])(?=.*[!?@#$%^&*_0-9]).{12,80}/)
```

Length is definitely the most important check, but testing for at least one capital, one lower, and one digit or special character is a pretty typical check. And it's easy enough to switch back to a length-only check!

Reply:
We can also keep both checks. But I remember reading somewhere that enforcing those characters does not necessarily help and can be annoying with people using password managers. Do you think it is still worth it nowadays?

Also, what do you think about this implementation:

```elixir
|> validate_length(:password, min: 12, max: 80)
|> validate_format(:password, ~r/[a-z]/, message: "at least one lower case character")
|> validate_format(:password, ~r/[A-Z]/, message: "at least one upper case character")
|> validate_format(:password, ~r/[!?@#$%^&*_0-9]/, message: "at least one digit or punctuation character")
```

Reply:
Length is 100% most important, and there is a diminishing return on additional requirements. I think it's worth it to enforce some variety -- I personally like length 12, and at least one digit or special character -- but it's backed by feelings over data, so I wouldn't consider it insecure if it was missing additional checks.

This implementation looks great since it's really easy to remove or update requirements individually.

Reply:
I will add the others as code comments then, so we provide some guidance, and people can  enforce what they prefer. :)

Reply:
"No dictionary words" might be debatable. Maybe "no words out of the top 10000 by usage of any dictionary" is more accurate. Maybe that is even easier to implement as it caps a black list if one should be used.

Reply:
It's already scoped by the [minimum length requirement](https://pages.nist.gov/800-63-3/sp800-63b.html#a3-complexity):

> Since user choice of passwords will also be governed by a minimum length requirement, this dictionary need only include entries meeting that requirement.

Reply:
As @danschultzer has mentioned, the new recommendations emphasize checks that are outside the scope of this generator - common (easy-to-guess) words, etc. Maybe we could add a comment pointing out that the checks here might not be sufficient and recommending the use of a dedicated password strength checker (frontend or backend).